Sources/Purchasing/Purchases/Purchases.swift

```diff
@@ -951,22 +950,27 @@ public extension Purchases {
         return await productsAsync(productIdentifiers)
     }
 
-    @objc(purchaseProduct:withCompletion:)
-    func purchase(product: StoreProduct, completion: @escaping PurchaseCompletedBlock) {
-        purchasesOrchestrator.purchase(product: product, package: nil, completion: completion)
+    @objc(purchaseProduct:metadata:withCompletion:)
+    func purchase(product: StoreProduct,
+                  metadata: [String: String]? = nil,
+                  completion: @escaping PurchaseCompletedBlock) {
+        purchasesOrchestrator.purchase(product: product, package: nil, metadata: metadata, completion: completion)
     }
 
-    func purchase(product: StoreProduct) async throws -> PurchaseResultData {
-        return try await purchaseAsync(product: product)
+    func purchase(product: StoreProduct, metadata: [String: String]? = nil) async throws -> PurchaseResultData {
+        return try await purchaseAsync(product: product, metadata: metadata)
```

Comment:
While adding a parameter with a default value is source compatible, it is not binary compatible. See [Library Evolution in Swift](https://www.swift.org/blog/library-evolution/): 

> ### Examples of non-resilient changes
> [...]
> - Adding a parameter to a function’s parameter list (even if a default value is provided)

This can cause crashes when other libraries depend on purchases-ios and encounter an incompatible version at runtime. 

Following [SemVer](https://semver.org/) strictly, this would require a major release.

Reply:
Reverting to your previous approach of adding new functions would be both source and binary compatible, but I definitely agree with @fire-at-will that that would result in too many purchase functions.